### ISSUE TITLE
Migrate to monorepo and update deps

### ISSUE THREAD

**Issue by DonsWayo** [foxifyjs/foxify#20]
As I like the project I want to contribute as much I can :)

**Is your feature request related to a problem? Please describe.**
Is more easy to handle multiple packages under one project, can build and watch and ever run in parallel without to link/portals and have several editors opens.

**Describe the solution you'd like**
I create a branch with a monorepo based structure with tubo https://turborepo.org/, include all the packages that foxify depends, update, make more standard builds and config (maybe missed something) and also builds commands for all at the same time.

**Describe alternatives you've considered**
Npm or yarn workspaces, but turbo repo have and extra layer that make more agile and quickly the development with cache and dependency management

**Additional context**
Maybe need to talk about some changes of build and also for benchmarks, also I update all dependencies and update some code.

I see that the current master is more slow than the npm published version (a half slow), some new implementation make more slow?

If this idea go ahead, later I propose to include and develop docs with docusaurus or similar to be more friendly for the developers.

Monorepo branch: https://github.com/DonsWayo/foxify/tree/feat/turbo-monorepo

**Comment by ardalanamini** [comment]
Hi @DonsWayo, Thanks for your interest in improving this project.

Firstly I'd like to point out that the idea of migrating to monorepo is very appealing to me especially due to the fact that at the moment at least I'm the sole maintainer.

But I still have some concerns about some parts of the way it is being implemented in many other open-source projects 🤔 (which might not be an issue at all, I haven't worked with monorepo in any open-source projects so far, so I'll just be mentioning my thoughts here so that we can come up with the optimal way of implementing it)

1. I'm starting to migrate to [pnpm](https://pnpm.io) instead of `npm` due to the way it manages the packages, in my opinion, it's a better solution for open-source projects (open-source developers are more likely to have many projects on their machines, I'm one of them 😄) so based on that and the way `pnpm` already handles workspaces, not sure if we need any extra tool 🤔 (just to point out that I still haven't looked at either of them deeply enough, so I might be wrong on this one 👨🏻‍💻)
2. The usual releasing process that I've seen in other projects is not exactly the way I think it should be, for example, I don't think it's reasonable to bump all package versions after changing just one of them, and in another example to bump the major on all when just some of them have breaking changes.
My current plan for the first stable version is to adopt a release pattern that'll allow incremental upgrading in case of introducing any breaking changes (which is always a painful process in big projects using a framework/library) after the `1.0.0` release, and that'll need more clear versioning and changelogs.

basically, the second point is my blocker for adopting monorepo here so far, Although I admit I haven't had the time to look for a solution as of yet.
Once this concern is resolved I'd be happy to move to use monorepo here 🚀

**Comment by ardalanamini** [comment]
@DonsWayo Also, thanks for the effort you put into implementing it so far, I hope we can get it in here fast enough 😅, it can make my life easier, for implementing my current roadmap to version `1.0.0`.

**Comment by DonsWayo** [comment]
Thank you for the amazing library!

I can tell you more about your points:

1 - use Pnpm

I agree and using pnpm is something that the monorepo can currently handle with some changes, I going to implement pnpm in the current branch.

The main reason of use turbo and not only the package manager workspaces is the layer of a cache of builds, build scripts, and builds orders...

If a monorepo have 10 packages, only build the project that makes the changes and not all again (detect file hashing changes)
Also, parallel works are easier without any other dependency between the OS system and many more things that you can check on the turbo website.

2 - use the same versions

I don't also agree to use the same versions, I think every package is different so don't need to be the same version at all, just every package has its version and is independent of others.

**Comment by ardalanamini** [comment]
Awesome, then let's find a good solution for the versioning and release first (especially in GitHub) and we'll start creating separated PRs for each implementation after that.

I think the first PR will be your migration to monorepo after the planning is done;
Then we'll move to pnpm and implement the release flow all in their related PRs.

I'll be creating the GitHub project and adding all the steps needed for this and the roadmap to first stable release 🤔🎉

**Comment by DonsWayo** [comment]
@ardalanamini 

- Update monorepo to PNPM
- Added CI build https://github.com/DonsWayo/foxify/actions/runs/3381635989 ✅

**Comment by ardalanamini** [comment]
I'm creating smaller issues for this enhancement, due to it being considerably big for review and containing some out of scope (which is moving to monorepo) changes.

For future reference let's first make the issues finalized and then start working on them and o avoid unnecessary time being put on them and over complicating the process.
I have created a GitHub project to keep track of the process.

I apologize for the inconvenience related this issue as the project is not exactly contributor friendly as much as I'd like it to be 😃
